Reinstate the Dhcp6Dxe driver veto, but only for QEMU

Backport of https://github.com/ipxe/ipxe/pull/1689 (commit 92fc305), which
closes https://github.com/ipxe/ipxe/issues/1688.

iPXE 5e06c866a removed the Dhcp6Dxe veto, but the edk2 bug it worked around
(https://github.com/tianocore/edk2/issues/10506, where EfiDhcp6Stop() spins
waiting for an event that never fires) is still unfixed. Booting over IPv6
under QEMU/OVMF therefore hangs at "iPXE initialising devices" before iPXE
reaches its banner.

The veto is restricted to a "QEMU" manufacturer because restoring it
unconditionally reintroduces https://github.com/ipxe/ipxe/issues/1358 on
Phoenix firmware, which is why 5e06c866a removed it in the first place.

Drop this patch once PR 1689 is merged and the pin includes it.

--- a/src/interface/efi/efi_veto.c
+++ b/src/interface/efi/efi_veto.c
@@ -480,6 +480,35 @@ efi_veto_vmware_uefipxebc ( EFI_DRIVER_BINDING_PROTOCOL *binding __unused,
 	return 1;
 }
 
+/**
+ * Veto QEMU Dhcp6Dxe driver
+ *
+ * @v binding		Driver binding protocol
+ * @v loaded		Loaded image protocol
+ * @v manufacturer	Manufacturer name, if present
+ * @v name		Driver name, if present
+ * @ret vetoed		Driver is to be vetoed
+ */
+static int efi_veto_qemu_dhcp6 ( EFI_DRIVER_BINDING_PROTOCOL *binding __unused,
+				 EFI_LOADED_IMAGE_PROTOCOL *loaded __unused,
+				 const char *manufacturer,
+				 const CHAR16 *name ) {
+	static const CHAR16 dhcp6[] = L"DHCP6 Protocol Driver";
+	static const char *qemu = "QEMU";
+
+	/* Check manufacturer and driver name */
+	if ( ! manufacturer )
+		return 0;
+	if ( ! name )
+		return 0;
+	if ( strcmp ( manufacturer, qemu) != 0 )
+		return 0;
+	if ( memcmp ( name, dhcp6, sizeof ( dhcp6 ) ) != 0 )
+		return 0;
+
+	return 1;
+}
+
 /** Driver vetoes */
 static struct efi_veto_candidate efi_vetoes[] = {
 	{
@@ -494,6 +523,10 @@ static struct efi_veto_candidate efi_vetoes[] = {
 		.name = "VMware UefiPxeBc",
 		.veto = efi_veto_vmware_uefipxebc,
 	},
+	{
+		.name = "QEMU Dhcp6",
+		.veto = efi_veto_qemu_dhcp6,
+	},
 };
 
 /**
